GNU patch 2.7.6 always needs a strip option

OrangeDog opened the pull request:
### What does this PR do?
When checking whether a patch applies in reverse, add the `-p0` option.
Without this, you get errors such as

>No file to patch.  Skipping patch.
>1 out of 1 hunk ignored
>can't find file to patch at input line 857
>Perhaps you should have used the -p or --strip option?

There may be additional bugs like this if the target is not a directory. The solution should be to default `strip=0`.

### What issues does this PR fix or reference?
Fixes: #52329

### Previous Behavior
Once a patch has been applied, all further runs fail.

### New Behavior
State now correctly detects whether the patch has been applied,

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

OrangeDog committed:
GNU patch 2.7.6 always needs a strip option

Fixes #52329
Without this, you get errors such as

>No file to patch.  Skipping patch.
>1 out of 1 hunk ignored
>can't find file to patch at input line 857
>Perhaps you should have used the -p or --strip option?

There may be additional bugs like this if the target is not a directory. The solution should be to default `strip=0`.

OrangeDog commented:
@terminalmage wrote this version of the state originally.
This will need tests.

cmcmarrow commented:
re-run all full

sagetherage commented:
@waynew this is a good one for the Test Clinic

twangboy reviewed (changes requested):
This will need a change log and a test

dwoz committed:
Add regression test and changelog for #52329

Adds test_patch_single_file_absolute_paths to pin the already-applied
detection behaviour for patches with absolute path headers. Also adds
changelog/57205.fixed.md per project convention.

Fixes #52329

dwoz committed:
Remove dead variable in test_patch_single_file_absolute_paths

patch_file was written to tmp_path but the state reads from state_tree;
only the patch_dest write is needed.
